### DISCUSSION TITLE
[Umbrella] improve Type Inference

### DISCUSSION THREAD

**Discussion by 0x53A** [fsharp/fslang-suggestions#1156] [General]
The type inference is one of the awesome parts of F#. Unfortunately, there are cases where it fails in (for someone unfamiliar with the exact spec and algorithm) _trivial_ cases. Even though most can be fixed with simple workarounds, this can be off-putting for a newcomer, and annoying for even a "veteran".

I would like to gather code samples here which you would expect to compile, but which don't, for whatever reason.

From this selection of currently failing code samples, we can then create PRs (if they are bugs), or more specific suggestions and RFCs.

/cc @matthid, I think you also mentioned a few failing cases in a previous issue, please add them here if you can remember them

**Comment by 0x53A** [comment]
### Type inference fails between Task and Task<'T>

note: this was worked around using inline constraints in the builder, so the actual issue was fixed, but it may be possible to improve inference with regards to subtyping. If you have other code-samples which exhibit the same behavior, please contribute them

original issue: https://github.com/Microsoft/visualfsharp/issues/3281

----------------


use this ``task`` builder: https://github.com/rspeele/TaskBuilder.fs/blob/6aa7ef6e47caf9f0af56413c4bb6d256763ffbaf/TaskBuilder.fs

and this F# code consuming the builder:

```F#
open System.Threading.Tasks
open FSharp.Control.Tasks.ContextSensitive

type X() =
    member this.Num = 0

let xTask : Task<X> = task { return X() }

let badTask =
    task {
        let! x = xTask
        return x.Num + 1
        //      ^ FS0072: Lookup on object of indeterminate type based on
        //        information prior to this program point.
    }
``` 

This caused the author the remove direct support for Task, and requires the user to wrap a non generic ``Task`` with a helper (``unitTask``), see https://github.com/rspeele/TaskBuilder.fs/commit/eeb42e52fce7d082dae1d25f2b297f9849b16f4b.

**Comment by 0x53A** [comment]
_(ToDo: Extract it to something debuggable)_

### Explicit type annotation is required on the left side of an assignment

see https://github.com/fsprojects/Paket/blob/master/src/Paket.Core/PaketConfigFiles/DependenciesFile.fs#L322

```F#
                    if Environment.GetEnvironmentVariable "PAKET_DEBUG_RUNTIME_DEPS" = "true" then
                        tracefn "Runtime dependencies: "
                        let (runtimeDepsFile:DependenciesFile) = DependenciesFile.FromSource ""
                        let (runtimeDepsFile:DependenciesFile) = runtimeDepsFile.AddFrameworkRestriction(Constants.MainDependencyGroup, getExplicitRestriction group.Options.Settings.FrameworkRestrictions)
                        let (runtimeDepsFile:DependenciesFile) = runtimeDepsFile.AddResolverStrategyForTransitives(Constants.MainDependencyGroup, group.Options.ResolverStrategyForTransitives)
                        let (runtimeDepsFile:DependenciesFile) = runtimeDepsFile.AddResolverStrategyForDirectDependencies(Constants.MainDependencyGroup, group.Options.ResolverStrategyForDirectDependencies)
                        let runtimeDepsFile =
                            runtimeResolutionDeps
                            |> Seq.fold (fun (deps:DependenciesFile) dep -> 
                                deps.AddAdditionalPackage(Constants.MainDependencyGroup, dep.Name, dep.VersionRequirement, dep.ResolverStrategyForTransitives, dep.Settings, false)) runtimeDepsFile

                        tracefn "Depsfile: \n%O" runtimeDepsFile
```

**Comment by 0x53A** [comment]
## Type solver fails to infer the method return value with regard to base-overloads taking callbacks

```F#
// Dummy, these are actually defined by Akka
[<AllowNullLiteral>]
type IUntypedActorContext = interface end

type ReceiveActor() =
    static member Context : IUntypedActorContext = null
    member x.Receive<'T>(cb:'T->bool) = ()
    member x.Receive(cb:'T->unit, ?pred:'T->bool) = ()

// my types

type IReplyChannel<'T> =
    abstract member ReplyWithValue : 'T -> unit
    abstract member ReplyWithException : exn -> unit

type InternalCallbackActorMessage<'T> =
| GetSelfInstance of IReplyChannel<CallbackActor<'T>>
| GetSelfContext of IReplyChannel<IUntypedActorContext>

and CallbackActor<'T>(cb:'T->unit) as self =
    inherit ReceiveActor()
    do base.Receive<'T>(cb)
    do base.Receive<InternalCallbackActorMessage<'T>>(self.OnInternalMsg)
    member self.OnInternalMsg(msg) =
        match msg with
        | GetSelfInstance rc -> rc.ReplyWithValue self
        | GetSelfContext rc -> rc.ReplyWithValue ReceiveActor.Context
```

__Actual:__

```
stdin(27,33): error FS0001: This expression was expected to have type
    'bool'    
but here has type
    'unit' 
```

__Workaround:__

Explicitly annotate the member method as returning unit:

``` member self.OnInternalMsg(msg) : unit = ```

__Reasoning:__

I think it is reasonable to expect this to work, because there is no uncertainty involved in the actual member implementation: it contains a match where both branches contain unit, so it itself must return unit. So it should short-circuit that. And as soon as it knows that the member method returns unit, it can also infer the correct base overload.

**Comment by 0x53A** [comment]
_Edited: thanks, updated! I still keep it as an example, because normally you would not expect to need to update the callsite when adding an optional parameter. But this is probably "wont-fix"._

### Applying tupled values to a constructor is inconsistent wrt optional parameters

_This is probably according to spec, but it may be worth to improve this, if it is possible without introducing ambiguity._

1) This works:

```F#
module Program

type T(x : int*int) = class end

let t = T(1, 2)
```

----------

2) If I then add a second, optional parameter, I need to also update the callsite:

```F#
type T(x : int*int, ?y : string) = class end

let t = T(1, 2) ❌
```

```
error FS0001: This expression was expected to have type
    'int * int'    
but here has type
    'int' 
```

Workaround: ``let t = T((1, 2))``  ✔

**Comment by dsyme** [comment]
@0x53A That's not the correct declaration for an optional parameter, try this
```
type T(x : int*int, ?y : string) = class end
```
when this works:
```
let t = T((1, 2))
```

**Comment by gsomix** [comment]
Can record type inference be improved in this case?

```fsharp
type Record1 = { id: int; title: string }
type Record2 = { id: int; name: string }

let test1 x = // OK
    let a = x.id
    let b = x.name
    ()
    
let test2 x =
    let a = x.id
    let b = x.title // ERR
    ()

let test3 x = // OK
    let b = x.title
    let a = x.id
    ()
```

> error FS: Type constraint mismatch. The type 'Record2' is not compatible with type 'Record1'

/cc @dsyme

**Comment by dsyme** [comment]
@gsomix Just use a type annotation in such a case, it's not that bad.

**Comment by gsomix** [comment]
@dsyme OK, but such behaviour may be confusing, especially for beginners. Can error message be improved instead?

Thanks!

/cc @forki

**Comment by dsyme** [comment]
> @dsyme OK, but such behaviour may be confusing, especially for beginners. Can error message be improved instead?

Definitely :)

**Comment by dsyme** [comment]
@0x53A re this:

> Explicit type annotation is required on the left side of an assignment

Could you give more detail as to why?

thanks

**Comment by cmeeren** [comment]
### let! binding in CE with extension overloads causes FS0072: Lookup on object of indeterminate type

Copied from https://github.com/Microsoft/visualfsharp/issues/4472. I don't think I have the necessary insight into how type inference works to identify the root cause and create a separate RFC on my own.

---

The following `AsyncResultBuilder` is a stripped-down version of the one in [cmeeren/Cvdm.ErrorHandling](https://github.com/cmeeren/Cvdm.ErrorHandling/).

```f#
type AsyncResultBuilder() =

  member __.Return (value: 'a) : Async<Result<'a, 'b>> =
    async { return Ok value }

  member __.Bind
      (asyncResult: Async<Result<'a, 'c>>,
       binder: 'a -> Async<Result<'b, 'c>>)
      : Async<Result<'b, 'c>> =
    async {
      let! result = asyncResult
      let bound =
        match result with
        | Ok x -> binder x
        | Error x -> async { return Error x }
      return! bound
    }

[<AutoOpen>]
module Extensions =

  // Having Async<_> members as extensions gives them lower priority in
  // overload resolution between Async<_> and Async<Result<_,_>>.
  type AsyncResultBuilder with

    member this.Bind
        (asnc: Async<'a>,
         binder: 'a -> Async<Result<'b, 'c>>)
        : Async<Result<'b, 'c>> =
      let asyncResult = async {
        let! x = asnc
        return Ok x
      }
      this.Bind(asyncResult, binder)


let asyncResult = AsyncResultBuilder()


let f () =
  asyncResult {
    let! str = asyncResult { return "" }
    return str.Length
           ^^^^^^^^^^
  }
```

#### Expected behavior

Everything compiles and works fine.

#### Actual behavior

I get the following error on `str.Length`:

`error FS0072: Lookup on object of indeterminate type based on information prior to this program point. A type annotation may be needed prior to this program point to constrain the type of the object. This may allow the lookup to be resolved.`

Note that intellisense is able to determine the type:

![image](https://user-images.githubusercontent.com/7766733/37151609-6f6a436e-22d6-11e8-8015-a7513b14e315.png)

#### Known workarounds

Add a type annotation:

```f#
let! (str: string) = asyncResult { return "" }
```

The problem also disappears if you remove the extension overload, but this of course has other undesired consequences.

**Comment by matthid** [comment]
> Could you give more detail as to why?

@dsyme @0x53A 
The minimal example is:

```fsharp

type Class(fileName) =
    member this.Resolve(s) =
        let c = Class.FromSource "s"
        let c = c.T1("s")
        let c = c.T2("s")
        let c = c.T3("s")
        ()
        
    member private c.T1(s) = c
    member private c.T2(s) = c
    member private c.T3(s) = c
    static member FromSource(source:string) : Class = Class(source)
```

Which doesn't compile with various `FS00072` errors.
I have found the following workarounds:

- either adding `: Class` to the `let c` expressions as shown in the example from @0x53A 
- Add `: Class` to the `member` declarations: `member private c.T1(s) : Class = c`
- Move the members above the `Resolve` method, which might lead to other code no longer working ;)

**Comment by cartermp** [comment]
Linking this: https://github.com/dotnet/fsharp/issues/7860

We've seen enough bug reports come in about not piping into generic functions needing type annotations that I think it's worth bringing up:

```fsharp
type CartItem = {
    ItemID: string
    ItemDescription: string
    ItemPrice: double
    Qty: int
}

type OrderHistoryItem = {
    SlipNumber: string
    OrderDate: DateTime
    MealTimeID: string
    PickupLocation: string
    ItemID: string
    ItemDescription: string
    ItemPrice: decimal
    Qty: int
}

let subtotalItems (items:CartItem list) =
    items
    |> List.groupBy (fun x -> (x.ItemID, x.ItemDescription, x.ItemPrice) )
    |> List.map (fun ((id,desc,price),itms)  ->
        {
            ItemID= id
            ItemDescription= desc
            ItemPrice= price
            Qty= List.sumBy (fun x -> x.Qty) itms // ERROR - 'itms' type mismatch
        })
```

I'd propose that this should not error anymore. It's true that piping makes this problem go away, but it's clearly not discoverable given the bug reports we get.

However, the same should _not_ apply to arbitrary members on classes. Although this would be a benefit to the user, the complexity of crawling type heirarchies should be something we're not doing here.

**Comment by Happypig375** [comment]
I'll add one: Overload resolution fails to determine the appropriate base constructor to call based on static member
```fs
type C() =
    inherit ResizeArray<int>(C.N():int) // Required!
    let t = ResizeArray<int>(C.N())
    member _.M() = ResizeArray<int>(C.N())
    static member N() = 4
```

**Comment by cmeeren** [comment]
I'd like to see better type inference for record copy-and-update expressions. When there are multiple matching record types in scope, it should default to the latest one, just like when using `.`.

```f#
type Record1 = { X: int; Y: string }
type Record2 = { X: int }

// Successfully infers Record2
let getX a = a.X 

// Fails with FS0667: Labels do not uniquely determine a corresponding record type.
// Ideally it should infer Record2.
let setX value a = { a with X = value }
```

**Reply by Xyncgas** [reply]
#1313 
Would be helpful for anonymous records to be inferred too

**Reply by cmeeren** [reply]
Not sure why you link to #1313; that doesn't seem related to my proposal.

**Reply by cmeeren** [reply]
I just created #1330.

**Reply by smoothdeveloper** [reply]
@cmeeren it is related to type inference (and everything or nothing is related, you pick 🙂)

**Reply by Xyncgas** [reply]
In your proposal you can simply ```let getX a = a.X ``` and a is something that has a 'field x', #1313 say you can do it for everything, for example if you have a record called myRecord that has a field X, a.X can be infered to myRecord.X, but it can infers to anonymous record that contains X if you didn't write a record called myRecord or if it's not the appropriate choice

Simply put, dynamic programming compiled. You can say
```
    let getX a =
        a.X |> ignore
        a.X1 |> ignore
        a.X1.X2 |> ignore
```
without having to define what a is but it's inferred and during compilation an anonymous record type would be created as {|X :{|X1:{|X2:'a|}|}|}

**Comment by dsyme** [comment]
I'm converting this to a discussion because so many things have been thrown into one thread, and most are not concrete proposals (separately - I checked one case list and it has been fixed and minmized that)

There's real value here about where type annotations are needed, but currently each one is by design and I'm not sure what the concrete proposal is except that people "want it to work".

**Comment by Xyncgas** [comment]
Honestly I would enjoy seeing type inference everywhere, to the point it feels like you are using reflections when you are writing codes, imo it is way more important than performance, of course we don't have to make it slow, but we can get conveniences of C# dynamic which we can say something is there to pretend it is so even though we are passing something to a function later we can use it now and that saves a lot wiring and doesn't have to make everything slower
